silkworm/sync/sentry_client.cpp

```diff
@@ -16,137 +16,300 @@
+    if (supported_protocol < 66) {
+        log::Critical(kLogTitle) << "remote sentry do not support eth/66 protocol, stopping...";
         stop();
         throw SentryClientException("SentryClient exception, cause: sentry do not support eth/66 protocol");
     }
 }
 
-void SentryClient::execution_loop() {
-    using RMS = rpc::ReceiveMessages::Scope;
+static sentry::common::Message sentry_message_from_outbound_message(const OutboundMessage& outbound_message) {
+    return sentry::common::Message{
+        sentry::eth::common_message_id_from_eth_id(outbound_message.eth_message_id()),
+        outbound_message.message_data(),
+    };
+}
+
+static SentryClient::PeerIds peer_ids_from_peer_keys(const silkworm::sentry::api::api_common::Service::PeerKeys& peer_keys) {
+    SentryClient::PeerIds peer_ids;
+    for (auto& peer_key : peer_keys) {
+        peer_ids.push_back(peer_key.serialized());
+    }
+    return peer_ids;
+}
+
+SentryClient::PeerIds SentryClient::send_message_by_id(const OutboundMessage& outbound_message, const PeerId& peer_id) {
+    auto message = sentry_message_from_outbound_message(outbound_message);
+    auto peer_public_key = sentry::common::EccPublicKey::deserialize(peer_id);
+    auto peer_keys = sync_spawn(tasks_, io_context_, sentry_client_->service()->send_message_by_id(std::move(message), std::move(peer_public_key)));
+    return peer_ids_from_peer_keys(peer_keys);
+}
+
+SentryClient::PeerIds SentryClient::send_message_to_random_peers(const OutboundMessage& outbound_message, size_t max_peers) {
+    auto message = sentry_message_from_outbound_message(outbound_message);
+    auto peer_keys = sync_spawn(tasks_, io_context_, sentry_client_->service()->send_message_to_random_peers(std::move(message), max_peers));
+    return peer_ids_from_peer_keys(peer_keys);
+}
+
+SentryClient::PeerIds SentryClient::send_message_to_all(const OutboundMessage& outbound_message) {
+    auto message = sentry_message_from_outbound_message(outbound_message);
+    auto peer_keys = sync_spawn(tasks_, io_context_, sentry_client_->service()->send_message_to_all(std::move(message)));
+    return peer_ids_from_peer_keys(peer_keys);
+}
+
+SentryClient::PeerIds SentryClient::send_message_by_min_block(const OutboundMessage& outbound_message, BlockNum /*min_block*/, size_t max_peers) {
+    auto message = sentry_message_from_outbound_message(outbound_message);
+    auto peer_keys = sync_spawn(tasks_, io_context_, sentry_client_->service()->send_message_by_min_block(std::move(message), max_peers));
+    return peer_ids_from_peer_keys(peer_keys);
+}
+
+void SentryClient::peer_min_block(const PeerId& peer_id, BlockNum /*min_block*/) {
+    auto peer_public_key = sentry::common::EccPublicKey::deserialize(peer_id);
+    sync_spawn(tasks_, io_context_, sentry_client_->service()->peer_min_block(std::move(peer_public_key)));
+}
 
+void SentryClient::execution_loop() {
```

Comment:
I would like to get rid of this loop. SentryClient could expose a method like "receive_messages(callback)" that is now possible having an async client. This method will spawn a coroutine that when a message is received will send it to subscribers.

Reply:
:+1: Let's try to improve this after https://github.com/torquem-ch/silkworm/pull/1008 . The signals/slots and sync methods could be removed if the calling code migrates to the async API, which I'm adding in #1008 for all required methods.